x-pack/plugins/security_solution/server/lib/detection_engine/signals/threat_mapping/create_threat_signals.ts

```diff
@@ -90,12 +91,45 @@ export const createThreatSignals = async ({
     perPage,
   });
 
+  const getMatchedThreats: GetMatchedThreats = async (ids) => {
+    const matchedThreatsFilter = {
+      query: {
+        bool: {
+          filter: {
+            ids: { values: ids },
+          },
+        },
+      },
+    };
+    const threatResponse = await getThreatList({
+      callCluster: services.callCluster,
+      exceptionItems: exceptionItems ?? [],
+      threatFilters: [...(threatFilters ?? []), matchedThreatsFilter],
+      query: threatQuery,
+      language: threatLanguage,
+      index: threatIndex,
+      listClient,
+      searchAfter: undefined,
+      sortField: undefined,
+      sortOrder: undefined,
+      logger,
+      buildRuleMessage,
+      perPage: undefined,
+    });
+
+    return threatResponse.hits.hits ?? [];
+  };
+
+  const threatEnrichment = (signals: SignalSearchResponse): Promise<SignalSearchResponse> =>
+    enrichSignalThreatMatches(signals, getMatchedThreats);
```

Comment:
Overall studying this, the one suggestion I could make is that sometimes functions within functions can be harder to maintain in the long run and makes reasoning (at least for me) about things a bit tricker. The maintenance can in some cases become messier as this function grows (as an example).

This is totally optional, but what I would do at this point with the code is make a new file:

```ts
get_matched_threats.ts
```

And then define within that file a function which takes all the extra parameters and dependencies and return your new function. That would make potential testing easier in the future.

```ts
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

import { ListClient } from '../../../../../../lists/server';
import {
  ThreatIndex,
  ThreatLanguageOrUndefined,
  ThreatQuery,
} from '../../../../../common/detection_engine/schemas/types';
import { ExceptionListItemSchema } from '../../../../../../lists/common';
import {
  AlertInstanceContext,
  AlertInstanceState,
  AlertServices,
} from '../../../../../../alerts/server';
import { getThreatList } from './get_threat_list';
import { ThreatListItem } from './types';
import { PartialFilter } from '../../types';
import { Logger } from '../../../../../../../../src/core/server';
import { BuildRuleMessage } from '../rule_messages';

export interface Options {
  services: AlertServices<AlertInstanceState, AlertInstanceContext, 'default'>;
  exceptionItems: ExceptionListItemSchema[];
  threatFilters: PartialFilter[];
  threatQuery: ThreatQuery;
  threatLanguage: ThreatLanguageOrUndefined;
  listClient: ListClient;
  logger: Logger;
  threatIndex: ThreatIndex;
  buildRuleMessage: BuildRuleMessage;
}

export const buildMatchedThreats = ({
  services,
  exceptionItems,
  threatFilters,
  threatQuery,
  threatLanguage,
  listClient,
  logger,
  threatIndex,
  buildRuleMessage,
}: Options): ((ids: string[]) => Promise<ThreatListItem[]>) => {
  return async (ids: string[]): Promise<ThreatListItem[]> => {
    const matchedThreatsFilter = {
      query: {
        bool: {
          filter: {
            ids: { values: ids },
          },
        },
      },
    };
    const threatResponse = await getThreatList({
      callCluster: services.callCluster,
      exceptionItems,
      threatFilters: [...threatFilters, matchedThreatsFilter],
      query: threatQuery,
      language: threatLanguage,
      index: threatIndex,
      listClient,
      searchAfter: undefined,
      sortField: undefined,
      sortOrder: undefined,
      logger,
      buildRuleMessage,
      perPage: undefined,
    });

    return threatResponse.hits.hits;
  };
};
```

Reply:
bd2a467